package/ota_daemon/src/ota_daemon.c

```diff
@@ -0,0 +1,535 @@
+  enum {
+    OPT_ID_DEBUG = 1,
+    OPT_ID_MODE_OTA,
+    OPT_ID_MODE_SETTINGS,
+    OPT_ID_URL,
+  };
+
+  // clang-format off
+  const struct option long_opts[] = {
+    {"debug",     no_argument,        0, OPT_ID_DEBUG},
+    {"ota",       no_argument,        0, OPT_ID_MODE_OTA},
+    {"settings",  no_argument,        0, OPT_ID_MODE_SETTINGS},
+    {"url",       optional_argument,  0, OPT_ID_URL},
+    {0, 0, 0, 0},
+  };
+  // clang-format on
+
+  int opt;
+
+  while ((opt = getopt_long(argc, argv, "", long_opts, NULL)) != -1) {
+    switch (opt) {
+    case OPT_ID_MODE_OTA: {
+      op_mode = OP_MODE_OTA_CLIENT;
+    } break;
+
+    case OPT_ID_MODE_SETTINGS: {
+      op_mode = OP_MODE_SETTINGS_DAEMON;
+    } break;
+
+    case OPT_ID_URL: {
+      opt_url = optarg;
+    } break;
+
+    case OPT_ID_DEBUG: {
+      opt_debug = true;
+    } break;
+
+    default: {
+      puts("Invalid option");
+      return -1;
+    } break;
+    }
+  }
+
+  if (op_mode == OP_MODE_OTA_CLIENT && opt_url == NULL) {
+    opt_url = OTA_DEFAULT_URL;
+    printf("Using default URL: %s\n", opt_url);
+  }
+
+  return 0;
+}
+
+static void ota_sig_handler(int signum, siginfo_t *info, void *ucontext)
+{
+  (void)ucontext;
+  piksi_log(LOG_DEBUG, "%s: received signal: %d, sender: %d", __FUNCTION__, signum, info->si_pid);
+  libnetwork_shutdown(NETWORK_TYPE_OTA);
+}
+
+static int configure_libnetwork(network_context_t *ctx, int fd, const char *url)
+{
+  network_status_t status = NETWORK_STATUS_SUCCESS;
+
+  if ((status = libnetwork_set_url(ctx, url)) != NETWORK_STATUS_SUCCESS) {
+    piksi_log(LOG_ERR, libnetwork_status_text(status));
+    return 1;
+  }
+
+  if ((status = libnetwork_set_fd(ctx, fd)) != NETWORK_STATUS_SUCCESS) {
+    piksi_log(LOG_ERR, libnetwork_status_text(status));
+    return 1;
+  }
+
+  if ((status = libnetwork_set_debug(ctx, opt_debug)) != NETWORK_STATUS_SUCCESS) {
+    piksi_log(LOG_ERR, libnetwork_status_text(status));
+    return 1;
+  }
+
+  status = libnetwork_set_continuous(ctx, false);
+  if (status != NETWORK_STATUS_SUCCESS) {
+    piksi_log(LOG_ERR, libnetwork_status_text(status));
+    return 1;
+  }
+
+  return 0;
+}
+
+static int ota_upgrade()
+{
+  int ret = 0;
+
+  const char *upgrade_cmd =
+    "sh -c 'set -o pipefail; "
+    "sudo upgrade_tool --debug " OTA_IMAGE_FILE_PATH
+    " && "
+    "sudo /etc/init.d/do_sbp_msg_reset'";
+
+  runit_config_t cfg = (runit_config_t){
+    .service_dir = RUNIT_SERVICE_DIR,
+    .service_name = "upgrade_tool",
+    .command_line = upgrade_cmd,
+    .finish_command = NULL,
+    .restart = false,
+  };
+
+  ret = start_runit_service(&cfg);
+
+  if (ret) {
+    piksi_log(LOG_ERR | LOG_SBP, "Failed to start upgrade_tool: %d", ret);
+  }
+
+  return ret;
+}
+
+static int ota_sha256sum(const char *expected)
+{
+  const char *cmd = "sha256sum";
+  char *const argv[3] = {"sha256sum", OTA_IMAGE_FILE_PATH, NULL};
+  char sha[65] = {0};
+  int ret = run_with_stdin_file(NULL, cmd, argv, sha, sizeof(sha));
```

Comment:
Just noticed while testing that `sha256sum` is left in zombie state after exit. It's easily missed as the upgrade does reboot. Indicates that child exit code is not handled correctly.

Reply:
Fixed by adding `wait(NULL)` for parent.

Reply:
👍